### ISSUE TITLE
High CPU usage when using version 254

### ISSUE THREAD

**Issue by jhedider** [jmxtrans/jmxtrans#441]
Hello,

I've been using JMXTrans to transfer some 301 metrics every minute.
At the start up everything works fine, CPU usage is about 15%. Then after a while CPU usage starts to climb progressively until reaching 95% constantly.
I checked with mission control the threads that are demanding CPU and found some Stormpot-pool1-thread-XX on the top of the list.

![jmxtrans_threads_consommant_cpu_1](https://cloud.githubusercontent.com/assets/18289818/14287981/3439d926-fb55-11e5-8f61-0b069913b8da.PNG)

I could correlate the CPU increase with exception of this type:

```
java.net.SocketException: Connection reset
        at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:113)
        at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
        at sun.nio.cs.StreamEncoder.writeBytes(StreamEncoder.java:221)
        at sun.nio.cs.StreamEncoder.implWrite(StreamEncoder.java:282)
        at sun.nio.cs.StreamEncoder.write(StreamEncoder.java:125)
        at java.io.OutputStreamWriter.write(OutputStreamWriter.java:207)
        at java.io.BufferedWriter.flushBuffer(BufferedWriter.java:129)
        at java.io.BufferedWriter.write(BufferedWriter.java:230)
        at java.io.Writer.write(Writer.java:157)
        at com.googlecode.jmxtrans.model.output.GraphiteWriter2.write(GraphiteWriter2.java:78)
        at com.googlecode.jmxtrans.model.output.support.WriterPoolOutputWriter.doWrite(WriterPoolOutputWriter.java:54)
        at com.googlecode.jmxtrans.model.output.support.ResultTransformerOutputWriter.doWrite(ResultTransformerOutputWriter.java:61)
        at com.googlecode.jmxtrans.jmx.ResultProcessor$1.run(ResultProcessor.java:58)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```
I assumed according to these facts that the high CPU usage is caused by the Stormpot writer pools.
I tried to change the code and include an invalidate statement on the writer that raised this exception as follows (class com.googlecode.jmxtrans.model.output.support.WriterPoolOutputWriter):
```
 public void doWrite(Server server, Query query, Iterable<Result> results)
    throws Exception
  {
    try
    {
      WriterPoolable writerPoolable = (WriterPoolable)this.writerPool.claim(this.poolClaimTimeout);
      try
      {
        this.target.write(writerPoolable.getWriter(), server, query, results);
      }
      catch (Exception ioe)
      {
        writerPoolable.invalidate();
        this.logger.warn("Invalidate WriterPoolable", ioe);
        throw ioe;
      }
      finally
      {
        writerPoolable.release();
      }
    }
    catch (InterruptedException e)
    {
      throw new IllegalStateException("Could not get writer from pool, please check if the server is available");
    }
  }
```
This change didn't help as I reproduced the CPU high usage.

According to the Stormpot documentation, the [expiration takes effect only after the release of the pooled object](http://chrisvest.github.io/stormpot/site/apidocs/index.html) and it has to be delegated to Stormpot itself by [defining an Expiration strategy](http://chrisvest.github.io/stormpot/tutorial.html).

I could not test more code changes as I got some other priority tasks to do. But I still need to fix this issue in order to use JMXTrans for monitoring. 
Is it possible to check this issue?

As a second question: is there a way not to use poolable writer as the pool contains always 1 instance and the writers are instantiated once per request as shows the class histogram bellow
![jmxtrans_nb_instances](https://cloud.githubusercontent.com/assets/18289818/14289360/bb42b4ec-fb5a-11e5-846c-12c4fc109b62.PNG)

Thanks in advance for your help and your support.

**Comment by gehel** [comment]
There was an issue with multiple writers being created instead of having a singleton writer. This should be fixed in #435. The release has not yet been made, but you can test the [latest SNAPSHOT](https://oss.sonatype.org/content/repositories/snapshots/org/jmxtrans/jmxtrans/255-SNAPSHOT/). Let me know if it works better and I'll create the release.

**Comment by jhedider** [comment]
Thanks for you quick answer.
I tried the snapshot of 28/03 already and it didn't help. I've just done the same test with jmxtrans-255-20160403.200657-7-all.jar and I still have the same issue. There are 1204 Stormpot-pool-3-thread-nnnn threads created (4 times the number of metrics knowing that for each metric we have 2 outputwriters configured).
At start up everything works fine but after several minutes one of the Stormpot-pool-3-thread-nnnn becomes a high CPU consumer then other similar thread use more and more CPU.
This is a new thread snapshot for the snapshot you adviced to test:
![jmxtrans_detail_threadsstormpot_20160406](https://cloud.githubusercontent.com/assets/18289818/14309581/a228cbbc-fbdd-11e5-8b04-7e44ce18add8.PNG)
Here's an example of what these threads are doing:
![jmxtrans_threaddetailsstormpot_20160406](https://cloud.githubusercontent.com/assets/18289818/14309624/ecb3eb30-fbdd-11e5-90f2-35fe03bb441f.PNG)
Normally each thread is dealing with different pool each pool has a size of 1 (default size as it is not specified on the json file).

We use only 2 writers because we have 2 different graphite receivers. I guess there should be only 2 writer pools?

**Comment by gehel** [comment]
This is strange and interesting. Yes, there should be only 1 pool for each writer (in your case, only 2 pools). Any chance you could send me a few thread dumps of when the problem occurs and your configuration?

**Comment by nodarret** [comment]
I have similar problem. 
For example i have Stormpot two threads in pool. First one is TIMED_WAITING but the other one after a while goes RUNNABLE with 100% CPU usage.
In production app for 6 Stormpot threads 3-4 stack in RUNNABLE using all CPU.

Example stack trace for same thread:

```
"Stormpot-pool-1-thread-2" - Thread t@29
   java.lang.Thread.State: RUNNABLE
	at java.util.concurrent.LinkedTransferQueue.unsplice(LinkedTransferQueue.java:1076)
	at java.util.concurrent.LinkedTransferQueue.awaitMatch(LinkedTransferQueue.java:715)
	at java.util.concurrent.LinkedTransferQueue.xfer(LinkedTransferQueue.java:647)
	at java.util.concurrent.LinkedTransferQueue.poll(LinkedTransferQueue.java:1273)
	at stormpot.BAllocThread.replenishPool(BAllocThread.java:104)
	at stormpot.BAllocThread.continuouslyReplenishPool(BAllocThread.java:86)
	at stormpot.BAllocThread.run(BAllocThread.java:78)
	at java.lang.Thread.run(Thread.java:745)

   Locked ownable synchronizers:
	- None

```
```
"Stormpot-pool-1-thread-2" - Thread t@29
   java.lang.Thread.State: RUNNABLE
	at java.util.concurrent.LinkedTransferQueue.xfer(LinkedTransferQueue.java:638)
	at java.util.concurrent.LinkedTransferQueue.offer(LinkedTransferQueue.java:1192)
	at stormpot.BAllocThread.proactivelyHealPoison(BAllocThread.java:165)
	at stormpot.BAllocThread.replenishPool(BAllocThread.java:119)
	at stormpot.BAllocThread.continuouslyReplenishPool(BAllocThread.java:86)
	at stormpot.BAllocThread.run(BAllocThread.java:78)
	at java.lang.Thread.run(Thread.java:745)

   Locked ownable synchronizers:
	- None
```
```
"Stormpot-pool-1-thread-2" - Thread t@29
   java.lang.Thread.State: RUNNABLE
	at stormpot.BAllocThread.replenishPool(BAllocThread.java:119)
	at stormpot.BAllocThread.continuouslyReplenishPool(BAllocThread.java:86)
	at stormpot.BAllocThread.run(BAllocThread.java:78)
	at java.lang.Thread.run(Thread.java:745)

   Locked ownable synchronizers:
	- None
```
Hope it helps somehow.

**Comment by kevinconaway** [comment]
I'm seeing the same thing.  My (un)-educated guess is that this method in the `BAllocThread` is getting stuck:


```java
  private void proactivelyHealPoison() {
    BSlot<T> slot = live.poll();
    if (slot != null) {
      if (slot.poison != null && (slot.isDead() || slot.live2dead())) {
        realloc(slot);
      } else {
        live.offer(slot);
      }
    }
  }
```

I'm guessing it can't "heal" that particular poisioned slot for some reason and it just continuously polls and offers the same slot item in a tight loop.

I'm trying to reproduce it locally but cannot thus far.

**Comment by kevinconaway** [comment]
@jhedider can you share how you were able to reproduce this locally?

**Comment by gehel** [comment]
My guess is the same. If for some reason we can't allocate a resource, stormpot will try again and again. We should either:

* stop using stormpot and go back to commons pool
* add a circuit breaker to the allocators, with a back off delay (we should anyway, whatever the pool)
* add a circuit breaker in stormpot itself, or talk with the authors to see if they have agood way to handle those kind of situations.

Appart from that, we should also ensure that only one pool is created (per backend), so in the worst case scenario, there should be only one thread going crazy.

**Comment by kevinconaway** [comment]
Maybe @chrisvest has some insight.

**Comment by gehel** [comment]
Most probably... @kevinconaway: you want to follow the point and open an issue on stormpot? Just to start the discussion there?

**Comment by jhedider** [comment]
I just would like to give you some elements from our side hoping it helps you diagnose and fix the issue.
We are now using version 249 with the same configuration and have a complete different behaviour:

- It creates between 60 to 90 threads VS. 1200 with the latest versions
- It uses about less than 5% of CPU VS. 15% constant CPU usage with the latest versions and strange CPU behaviour when some exceptions happens.

I agree with the post of gehel saying that Stormpot threads turn crazy looping when something wrong happens with the instances in the pool. I agree also that only one pool instance should be created by writer type.

**Comment by chrisvest** [comment]
Hi, author of Stormpot here.

I have a couple of questions:

1. As I understand, there is supposed to be only two pools created, because there are only 2 Graphite instances in this case. Stormpot starts 1 thread per pool instance, which runs until the pool is shut down. Observing 60 to 90 (or 1200) Stormpot threads indicates that more than 2 pools are being created. Where are these extra pools coming from? Is there a process somewhere that churns through pools in response to something?
2. Stormpot is failing to heal the poisoned slot because the allocator is either returning `null` or throwing an exception. Probably the latter. What does the stack trace of such an exception look like?

We might be able to work around the problem if we find out why the threads start spinning like this.

Meanwhile, it looks like there's also a real bug in Stormpot: Apparently the pool will obsess over any poisoned slot it gets so much, that it won't notice if you ask it to shut down. Clearly there's no point in trying to reallocate a poisoned slot if we are about to shut down anyway, so the priorities should be the other way around.

**Comment by gehel** [comment]
1) there is a bug in jmxtrans where we do not consolidate the writers instances correctly, and we do create many more instances than we should. The previous implementation based on common pools did hide the issue by sharing a single pool between all instances and having the output writer class being extremely lightweight. This is mostly fixed in latest snapshot, but not yet tested enough.

2) I'm not completely sure on the analysis, but my understanding is: we have an allocator allocating TCP connections to a remote server and binding that connection. If that remote server is down, the allocation will always fail, until the remote server is up again. Short transient allocation errors are handled well by re allocating until the pool is full again, but this does not work as well for longer allocation issues.

Again, I'm not 100% sure of my reading of this issue, but that's my best educated guess...

**Comment by chrisvest** [comment]
Okay. So some kind of back-off would be nice, if allocation keeps failing after a short initial burst of reallocation attempts.

**Comment by gehel** [comment]
A circuit breaker, with a closed, open, half-closed state would be great. That would be fairly easy to implement in the allocator itself, but ineffective if I read the contract correctly. Stormpot would have no way to differentiate between a slow allocator and an open circuit breaker. I'm not familiar enough with stormpot to suggest an implementation on that side...

**Comment by chrisvest** [comment]
It's a bit of a hack, but for Stormpot 2.x you can do the circuit breaker in the allocator, and extend the time between retries by sleeping. The problem with this is that there's only one allocation thread, so if you make it sleep then it won't be able to do anything else. In 3.0 there will be more than one thread.

> Meanwhile, it looks like there's also a real bug in Stormpot: Apparently the pool will obsess over any
> poisoned slot it gets so much, that it won't notice if you ask it to shut down.

Turns out I was wrong about this part. The pool will shut down fine when asked, regardless of how many exceptions it has collected. Not having any back-off strategy is still an issue, though.

**Comment by kevinconaway** [comment]
Chris, apart form backing off of the reallocation, it seems like like the poision healing process should take in to account the expiration policy for the pool.

We are defining an expiration policy for the pool:

```java

	@Override
	public boolean hasExpired(SlotInfo<? extends SocketPoolable> info) throws Exception {
		Socket socket = info.getPoolable().getSocket();
		try {
			return socket == null
					|| !socket.isConnected()
					|| !socket.isBound()
					|| socket.isClosed()
					|| socket.isInputShutdown()
					|| socket.isOutputShutdown();
		} catch (Exception e) {
			log.warn("Socket {} is expired", socket, e);
			return true;
		}
	}
```

I think the "healing" process should also check if the slot is expired and if so, discard it instead of trying to continuously heal it.  

Thoughts?

**Comment by kevinconaway** [comment]
@gehel I'm testing locally and I'm seeing the OutputWriters initialized twice.  Below is the test config I'm using

```
{
  "servers": [{
    "port": "11061",
    "host": "localhost",
    "numQueryThreads": 2,
    "outputWriters": [{
      "@class": "com.googlecode.jmxtrans.model.output.GraphiteWriterFactory",
      "rootPrefix": "jmxtrans.test.zookeeper",
      "port": "2003",
      "host": "127.0.0.1"
    }],
    "queries": [{
      "obj": "java.lang:type=OperatingSystem",
      "attr": [
        "SystemLoadAverage",
        "AvailableProcessors",
        "TotalPhysicalMemorySize",
        "FreePhysicalMemorySize",
        "TotalSwapSpaceSize",
        "FreeSwapSpaceSize",
        "OpenFileDescriptorCount",
        "MaxFileDescriptorCount"
      ]
    },
      {
        "obj": "java.lang:type=Memory",
        "attr": [
          "HeapMemoryUsage",
          "NonHeapMemoryUsage"
        ]
      },
      {
        "obj": "java.lang:type=GarbageCollector,name=*",
        "attr": [
          "CollectionCount",
          "CollectionTime"
        ]
      },
      {
        "obj": "java.lang:type=Threading",
        "attr": [
          "DaemonThreadCount",
          "PeakThreadCount",
          "ThreadCount",
          "TotalStartedThreadCount"
        ]
      }
    ]
  }]
}
```

The OutputWriters first get created when the config JSON gets deserialized in the Server constructor:

https://github.com/jmxtrans/jmxtrans/blob/master/jmxtrans-core/src/main/java/com/googlecode/jmxtrans/model/Server.java#L215

The second time is happens is shortly after that the ServerListBuilder builds the server list by copying the outputwriters from the parsed model:

https://github.com/jmxtrans/jmxtrans/blob/master/jmxtrans-core/src/main/java/com/googlecode/jmxtrans/ServerListBuilder.java#L65

The `SingletonOutputWriterFactory` also creates an OutputWriter from the factory.

This seems incorrect to me.

**Comment by gehel** [comment]
@kevinconaway yes, this is most probably wrong, but also most probably of no consequence. There is a lot of ugly code there to keep backward compatibility. The Server instance created when parsing JSON configuration files is actually discarded once they are rewritten by the ServerListBuilder. At some point, we should stop trying to have a the classes used to parse configuration and the classes used at runtime be the same. That would mean more code, but much easier to read.

About your comment on stormpot poison healing, my reading of the code (yes, the code is always right, but it is misleading as well) is a bit different. The slots are poisoned because they expire, healing the slots means reallocating new poolable objects. What might be misleading, is that contrary to other pools, stormpot allocates a constant number of objects and asynchronously tries to ensure that the full pool size is correctly allocated and ready to be used. (@chrivest might confirm this).

**Comment by kevinconaway** [comment]
> yes, this is most probably wrong, but also most probably of no consequence

Are you sure?  It will create one pool per outputwriter.  If users specify a writer per query (as seems to be common practice), this will result in a lot of extra object pools and connections to the underlying systems.


>What might be misleading, is that contrary to other pools, stormpot allocates a constant number of objects and asynchronously tries to ensure that the full pool size is correctly allocated and ready to be used. 

If thats true, is that the model that you want for connections to ephemeral systems that may go up/down?

**Comment by gehel** [comment]
> > yes, this is most probably wrong, but also most probably of no consequence

> Are you sure? It will create one pool per outputwriter. If users specify a writer per query (as seems to be common practice), this will result in a lot of extra object pools and connections to the underlying systems

Actually, you're right, those unneeded pools are not shutdown and not garbage collected. (Yeah, that part of the code is crap).

> > What might be misleading, is that contrary to other pools, stormpot allocates a constant number of objects and asynchronously tries to ensure that the full pool size is correctly allocated and ready to be used.

> If thats true, is that the model that you want for connections to ephemeral systems that may go up/down?

Yes it is, if there is a circuit breaker ensuring that we don't send a storm of reconnect. And actually, this circuit breaker requirement is also true (to a lesser extent) for a resizeable, on demand pool. It just leave more options as to where to implement a circuit breaker.

**Comment by kevinconaway** [comment]
> Actually, you're right, those unneeded pools are not shutdown and not garbage collected.

Seems like a cheap fix would be to change `ConfigurationParser#parseServers` to do away with the ServerListBuilder.  I tested the following snippet locally via a test configuration file as well as the int/unit test:

```
	public ImmutableList parseServers(Iterable<File> jsonFiles, boolean continueOnJsonError) throws LifecycleException {
		ImmutableList<Server> servers = ImmutableList.of();
		for (File jsonFile : jsonFiles) {
			try {
				JmxProcess process = jsonUtils.parseProcess(jsonFile);
				log.debug("Loaded file: {}", jsonFile.getAbsolutePath());
				servers = ImmutableList.copyOf(process.getServers());
			} catch (Exception ex) {
				String message = "Error parsing json: " + jsonFile;
				// error parsing one file should not prevent the startup of JMXTrans
				if (continueOnJsonError) log.error(message, ex);
				else throw new LifecycleException(message, ex);
			}
		}
		return servers;
	}
```

Do you want me to submit a PR for that?

**Comment by gehel** [comment]
If I read that correctly, you only return the servers for the last file that is read. It's late here, so my eyes are probably crossed. If you have a PR and unit tests showing that it works, please send them! In any case, I'm always thrilled to receive PRs... (btw, you seem to invest a fair amount of time in jmxtrans, send me an email if you want to join the team).

**Comment by chrisvest** [comment]
@kevinconaway 
> Chris, apart form backing off of the reallocation, it seems like like the poision healing process should
> take in to account the expiration policy for the pool.
> I think the "healing" process should also check if the slot is expired and if so, discard it instead of trying
> to continuously heal it.

A poisoned slot is one that failed to get an allocated poolable object, so there is nothing to check the expiration of. Slots are sent to reallocation because they either have poison (always a failed allocation) or their associated object has expired.

**Comment by kevinconaway** [comment]
@gehel you're right.  I created #453 to track this issue separately so as not to derail this one where the the main focus should be stormpot IMO

**Comment by gehel** [comment]
Well... creating duplicate OutputWriters creates so many issues that optimizing the stormpot allocator makes almost no sense before we fix them. Thanks for creating the issue, it makes the problem more visible!

Now I (or someone else) just need to find time to clean this mess...

**Comment by kevinconaway** [comment]
Any update here?  254 is unusable due to this issue.  Can we revert back to using commons-pool (or something else) until stormpot is fixed?

**Comment by gehel** [comment]
We should fix it on our side. It's not that hard to write a circuit breaker allocator. I just need to find the time... It's probably just as much effort than switching back to common-pools.

In the meantime, the old output writers are still using common-pools. I should write some docs to that regard...

**Comment by kevinconaway** [comment]
> The problem with this is that there's only one allocation thread, so if you make it sleep then it won't be able to do anything else.

Does imply that the reallocator should retry for a period of time and then stop to the stormpot thread do something else?

**Comment by gehel** [comment]
The allocator thread is doing only allocations (as far as I understand the code). So blocking it is exactly what we want. We'll stop the allocation. At the moment stormpot does not have a multithreaded allocator, so there isn't even an issue of having to take care of that. We still need to keep some history of failures to take good decision on when to block, but that should not be too much of an issue.

**Comment by kevinconaway** [comment]
> So blocking it is exactly what we want. 

Is there a potential to starve out other slots from being allocated though?  Lets say there are two socket poolables, one to graphite, one to statsd and both become poisoned.  The graphite poolable is poisoned because the server disappeared.  The statsd poolable was poisioned due to a network blip.

If we start healing the graphite poolable, the reallocation process could spin forever until the server comes back up.  If we stop it after some period, the statds poolable would then have a chance at getting healed.

Thoughts?

**Comment by gehel** [comment]
We have different pools for different output writers. And each pool has its own allocator thread. So all slots in a single pool are all equivalents. That was actually one of the reason to move to stormpot instead of having a global keyed pool.

**Comment by gehel** [comment]
I added a few comments on your commit. Not entirely sure I understand stormpot correctly, but read my comments and see if they make sense...

**Comment by gehel** [comment]
Thanks a lot for that PR! I had in mind something closer to a circuit breaker, but your implementation with the retryer seems just as good.

And send me an email if you want in. JmxTrans is looking for talents!

**Comment by kevinconaway** [comment]
I tested the changes after #458 and #453 were merged and I was able to recreate the issue of the CPU spiking.  Here is what happens:

I started jmxtrans and pointed to my local graphite server.  After letting it run for a bit, I restarted the graphite server.  In the jmxtrans log, I saw a broken pipe exception:

```
12 May 2016 13:41:19] [jmxtrans-result-1] 28321861 WARN  (com.googlecode.jmxtrans.jmx.ResultProcessor$1:60) - Could not write results <omitted>  to output writer com.googlecode.jmxtrans.model.output.support.ResultTransformerOutputWriter@44be647b
java.net.SocketException: Broken pipe
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at sun.nio.cs.StreamEncoder.writeBytes(StreamEncoder.java:221)
	at sun.nio.cs.StreamEncoder.implWrite(StreamEncoder.java:282)
	at sun.nio.cs.StreamEncoder.write(StreamEncoder.java:125)
	at java.io.OutputStreamWriter.write(OutputStreamWriter.java:207)
	at java.io.BufferedWriter.flushBuffer(BufferedWriter.java:129)
	at java.io.BufferedWriter.write(BufferedWriter.java:230)
	at java.io.Writer.write(Writer.java:157)
	at com.googlecode.jmxtrans.model.output.GraphiteWriter2.write(GraphiteWriter2.java:77)
...snip...
```
When this happens we will explicitly expire the slot:
https://github.com/jmxtrans/jmxtrans/blob/master/jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/support/WriterPoolOutputWriter.java#L55

Which increments the number of poisoned slots:
https://github.com/chrisvest/stormpot/blob/master/src/main/java/stormpot/BSlot.java#L274

Stormpot notices that the slot is now poisioned and attempts to reallocate it:
https://github.com/chrisvest/stormpot/blob/master/src/main/java/stormpot/BAllocThread.java#L263

In #realloc, it sets the `slot.poison=null` and successfully reallocates the slot.  However it does *not* decrement the number of poisoned slots.

Thus, the `proactivelyHealPoison` again goes in to an infinite loop because the number of poisoned slots is still 1 but the slot is no longer poisoned

@chrisvest thoughts here?

**Comment by chrisvest** [comment]
Yeah, that's a bug. Good analysis. I'll fix it in the 2.4 branch when I have a spare moment.

**Comment by chrisvest** [comment]
@kevinconaway Please try with a snapshot build of this branch: https://github.com/chrisvest/stormpot/tree/2.4.1

I think the bug can still manifest through an unlucky interaction with shrinking the pool. I think that's a lot less likely, so I'll look at it later when I have more time.

**Comment by kevinconaway** [comment]
Thanks @chrisvest, I tested it out and that fixed the issue.  Will you be able to release this so that @gehel can incorporate it in to a build for jmxtrans?

That said, I still see the CPU usage around 1-2%.  I'm guessing that this is because the `BAllocThread` is continuously spinning every 50ms in replenishPool:
https://github.com/chrisvest/stormpot/blob/master/src/main/java/stormpot/BAllocThread.java#L101

Does it make sense to use something like wait/notify (or the higher level concurrency primitives) so that the thread only does work when it has to?

**Comment by chrisvest** [comment]
It has multiple tasks, so it wakes up periodically to check on all of them. This determines the timeliness of background expiration checking and of pool size changes, at least. It could be made tuneable, but it's a diminishing return.

**Comment by kevinconaway** [comment]
Got it.  I suppose you couldn't proactively perform background expiration without waking up. Resizing and healing poisoned slots could be done via wait/notify, I think.

Is there anything else that needs to be done before releasing a 2.4.1 version or any other testing I can assist with¿

**Comment by chrisvest** [comment]
Precise sleeping - your wait/notify suggestion - is too complicated and risky to tackle in a 2.4.1 release. And I'm redoing the background task scheduling in 3.0 anyway. 

I found 2 other cases where it's using more CPU than it strictly needs. I want to fix those as well.

**Comment by kevinconaway** [comment]
Hi @chrisvest were you able to fix those remaining issues?

**Comment by chrisvest** [comment]
@kevinconaway Still working on it. The fixes introduced timeliness problems in other areas, and this needs fixing as well.

**Comment by kevinconaway** [comment]
Hey @chrisvest, any ETA on when a fixed version will be ready to integrate here?

**Comment by chrisvest** [comment]
@kevinconaway I don't have much spare time to work on it right now, but it will be out soon. I've had to work through crashing JVMs on TravisCI, and my Maven signing key expiring.

**Comment by gehel** [comment]
@kevinconaway we coudl just go back to commons-pool in the meantime...

**Comment by chrisvest** [comment]
I have just released Stormpot 2.4.1 that I hope will fix the issue of the 100% CPU usage.

**Comment by gehel** [comment]
@chrisvest thanks a bunch! We'll let you know...

**Comment by gehel** [comment]
I updated master to stormpot 2.4.1 (#468). @kevinconaway any chance you could test and let us know if that works better?

Thanks!

**Comment by kevinconaway** [comment]
The high CPU usage issue is fixed.  However, as [before](https://github.com/jmxtrans/jmxtrans/issues/441#issuecomment-219044162),  I'm still seeing CPU usage of around 0.7-1.8% on my 8 core VM that I'm testing with.

Taking a few thread dumps puts the Stormpot threads here:

```java
BSlot<T> slot = dead.poll(deadPollTimeout, TimeUnit.MILLISECONDS);
```

**Comment by gehel** [comment]
@kevinconaway so, do you think it is worth going back to commons-pool? @chrisvest any idea on a timeline to fix this CPU usage issue?

**Comment by kevinconaway** [comment]
{quote}
so, do you think it is worth going back to commons-pool?
{quote}

Hard to say, what was the impetus for moving to stormpot in the first place?

I think my preference now would be to release the fixed version ASAP because the current version is plainly unusable.  

Evaluating moving back to commons-pool could be deferred until later or perhaps stormpot 3.0 will be out by then with more of these issues fixed.

**Comment by chrisvest** [comment]
Removing the periodic checking will take a long time. It would be a side-effect of changes in 3.0, which is still far from ready.

**Comment by gehel** [comment]
Release 256 create...

**Comment by etotheipi** [comment]
I'm seeing this issue in jmxtrans-259.  Is there a workaround for this?  Is there a previous version I can switch to that was pre-stormpot?  This worked for a brief period of time before it went berserk and filled 4 GB of log files overnight with the stormpot errors.  

```[27 Sep 2016 11:05:52] [Stormpot-pool-3-thread-2] 62443047 ERROR (com.googlecode.jmxtrans.model.output.support.pool.RetryingAllocator$1:77) - Error allocating slot
java.net.ConnectException: Connection refused
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at com.googlecode.jmxtrans.model.output.support.pool.SocketAllocator.allocate(SocketAllocator.java:61)
	at com.googlecode.jmxtrans.model.output.support.pool.SocketAllocator.allocate(SocketAllocator.java:37)
	at com.googlecode.jmxtrans.model.output.support.pool.RetryingAllocator$2.call(RetryingAllocator.java:89)
	at com.googlecode.jmxtrans.model.output.support.pool.RetryingAllocator$2.call(RetryingAllocator.java:86)
	at com.github.rholder.retry.AttemptTimeLimiters$NoAttemptTimeLimit.call(AttemptTimeLimiters.java:78)
	at com.github.rholder.retry.Retryer.call(Retryer.java:160)
	at com.googlecode.jmxtrans.model.output.support.pool.RetryingAllocator.allocate(RetryingAllocator.java:86)
	at stormpot.ReallocatingAdaptor.allocate(ReallocatingAdaptor.java:37)
	at stormpot.BAllocThread.alloc(BAllocThread.java:225)
	at stormpot.BAllocThread.realloc(BAllocThread.java:295)
	at stormpot.BAllocThread.proactivelyHealPoison(BAllocThread.java:167)
	at stormpot.BAllocThread.replenishPool(BAllocThread.java:123)
	at stormpot.BAllocThread.continuouslyReplenishPool(BAllocThread.java:88)
	at stormpot.BAllocThread.run(BAllocThread.java:80)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by chrisvest** [comment]
Stormpot 3.1 was just released. It introduces an operating mode where there's no background thread. This means the background expiration checking and the background healing of failed allocations, are disabled in that mode. That should eliminate the constant low CPU usage that's been observed. Pools using this operating mode are created with `Pool.fromInline(...)`. Note that 3.x versions of Stormpot require Java 11.
